### PR TITLE
V2 role based access

### PR DESCRIPTION
**Pull request by sijumoncy**
#246 #247 #258 

This PR contains the Authentication and access rule check related with Role Based Access System implementation. As we discussed the things are implemented. The test cases are added separately for POST PUT and GET with different possibilities and get is also filtered with roles and apps. Some issues are caused in the other tests due to the auth logic and all auth implemented end points (Meta Content and source) test are working properly.

### CODE REVIEW

File: app/routers/content_apis.py [hunk 1]
```diff
@@ -12,6 +13,13 @@
 router = APIRouter()
 auth_handler = AuthHandler()
 
+#optional authentication with token or none
+optional_oauth2_scheme = OAuth2PasswordBearer(tokenUrl="auth", auto_error=False)
+async def get_user_or_none(token: str = Depends(optional_oauth2_scheme)):
+    """optional auth for getting token of logined user not raise error if no auth"""
+    # print("token===>",token)
```
Comment (kavitharaju) [inline app/routers/content_apis.py:20]: I assume this is creating a dummy user token with None values, in the absence of actual token in the header. This looks like a good way of handling the `noauthrequired` cases. But a few thoughts about it.

As other endpoints defined else where also might need this, may be we can move this definition to a common place like `authentication.py` and import to different sub modules in routers is required.

Another thought is this. Even though we are specifying different access rights associated with different endpoints in the  `api_permission_map.py` the authetication part is now specified at the endpoint definition in the router module itself. We have 3 ways we do it now. 
1. For endpoints that dont need an auth, we omit the Dependancy
2. For those that need an auth we use  `Depends(auth_handler.kratos_session_validation)`
3. For those end points that has optional auth requirement we use, `Depends(get_user_or_none)`
Can combine all these into one function and call that same function uniformly as dependancy of every endpoint? Let it be this `get_user_or_none()` function. In the access check decorator,  we can raise error or accept the None user, based on the APIs permission map. That way all the endpoint specific auth specifying will be required only at that one file.

File: app/crud/structurals_crud.py [hunk 2]
```diff
@@ -291,7 +291,7 @@ def create_source(db_: Session, source: schemas.SourceCreate, source_name, user_
         metaData = source.metaData,
```
Comment (kavitharaju) [inline app/crud/structurals_crud.py:291]: Doesn't this part(both in create source and update source methods) require modification to store the access tags specified by content uploader in the DB?

File: app/authentication.py [hunk 3]
```diff
@@ -283,26 +326,36 @@ async def wrapper(*args, **kwargs):
...
-                    raise PermisionException("Access Permission Denied for the URL")
-                db_.commit()#pylint: disable=E1101
-                db_.refresh(db_resource)#pylint: disable=E1101
-            elif required_params['request_context']['method'] == 'GET':#pylint: disable=E1126
-                db_resource = response
-                verified = check_access_rights(db_, required_params, db_resource)
-                if not verified:
-                    raise PermisionException("Access Permission Denied for the URL")
+            if len(response) > 0:
+                if required_params['request_context']['method'] != 'GET':#pylint: disable=E1126
```
Comment (kavitharaju) [inline app/authentication.py:339]: I assume this case is added for update or create functions. In that case, when the response in first obtained in this decorator before the `db.commit()`, please make sure the final response include `id` values also.

Comment (sijumoncy) [reply app/authentication.py:339]: fixed id is coming

File: app/routers/content_apis.py (hunk 1 above)
Comment (sijumoncy) [reply app/routers/content_apis.py:20]: work on this idea to combine auth function

**Pull request merged**